man/emu-push.1
.TH emu-push 1 July 14, 2015 "version 0.2.0" "Emu Manual"
.SH NAME
emu\-push \- record source snapshots
.SH SYNOPSIS
.B emu push
[<options>] [<sink> ...]
.SH DESCRIPTION
Push a new snapshot of the source directory onto one or more
sinks. If none are specified, push to all sinks.
.PP
The difference between the source and a sink's HEAD is transferred
and linked against the existing content, combining to produce a full
snapshot of the source directory. This enables emu sinks to be very
compact, allowing multiple full snapshots to be stored in a space
little larger than a single full snapshot.
.PP
Some of the features of snapshot creation are:
.IP \[bu]
supports copying links, devices, owners, groups, and permissions
.IP \[bu]
exclude and exclude\-from options
.IP \[bu]
does not require super\-user privileges
.PP
Emu uses the
.B rsync
(1) file\-copying tool for its delta\-transfer
algorithm, which reduces the amount of data sent by sending only the
differences between the source files and the existing files in the
sink. Rsync finds files that need to be transferred using a "quick
check" algorithm that looks for files that have changed in size or in
last\-modified time.
.SH OPTIONS
\-d \-\-dry-run
Perform a trial run with no changes made.
.TP
\-f \-\-force
Force push the snapshot, ignoring any concurrency locks on the source
or sink.
.TP
\-i \-\-ignore-errors
Ignore errors in the file transfer stage. Errors in the file transfer
stage can be caused by insufficient permissions to transfer file
attributes. Using this argument causes the push to continue despite
any such errros.
.TP
\-\-no-archive
Do not preserve file attributes such as ownership and permission bits
when creating the snapshot.
.TP
\-\-no-owner
Do not preserve file attributes such as ownership and permission bits
when creating the snapshot.
.TP
\-v \-\-verbose
Increase verbosity.
.TP
\-\-version
Show version information and quit.
.TP
\-h \-\-help
Show this help and quit.
.SH EMU
Part of the
.B emu
(1)
suite
.SH AUTHOR
Chris Cummins <chrisc.101@gmail.com>
.SH SEE ALSO
.B emu
(1)
.B rsync
(1)


CHANGELOG
emu-0.0.1 (Thu Jan 17 11:51:54 GMT 2013)
--------------------------------------------------------------------------------

Date:   Thu Jan 17 11:51:32 2013 +0000
    Rename make-release mkrelease

Date:   Thu Jan 17 11:50:12 2013 +0000
    make: remove release rule

Date:   Thu Jan 17 11:47:45 2013 +0000
    Add release documentation

Date:   Thu Jan 17 11:32:08 2013 +0000
    libemu: exit code vars

Date:   Thu Jan 17 11:27:21 2013 +0000
    libemu: update documentation

Date:   Thu Jan 17 11:15:07 2013 +0000
    make: Don't use $(QUIET) without good cause

Date:   Thu Jan 17 11:12:20 2013 +0000
    make: use '-' to separate words in rules

Date:   Thu Jan 17 11:10:30 2013 +0000
    test: Fix libemu var name

Date:   Mon Dec 17 16:11:32 2012 +0000
    make: Fix typo

Date:   Mon Dec 17 16:11:04 2012 +0000
    make: Rename 'default' to 'help'

Date:   Mon Dec 17 16:10:23 2012 +0000
    make: Remove 'VERSION' var

Date:   Mon Dec 17 16:09:37 2012 +0000
    make: Use portability variables

Date:   Mon Dec 17 16:07:16 2012 +0000
    make: Assign phony rules

Date:   Mon Dec 17 16:06:33 2012 +0000
    Use conventional $(QUIET) Makefile assignment

    - make V=1 will now be verbose

Date:   Thu Dec 6 20:56:11 2012 +0000
    libemu: Refactor command_not_found

Date:   Thu Dec 6 20:54:59 2012 +0000
    Add global constant $EMU_VERSION

Date:   Thu Dec 6 20:53:40 2012 +0000
    libemu: Correct echo usage

Date:   Thu Dec 6 20:50:10 2012 +0000
    test: Add libemu echo test

Date:   Thu Dec 6 20:42:05 2012 +0000
    Remove $PROG variable

    - Refactor '$PROG' -> 'emu'
Merge: 2b4a93e 53ac6b2

Date:   Thu Dec 6 20:37:48 2012 +0000
    Merge branch 'wip/echo' into next

Date:   Thu Dec 6 20:36:49 2012 +0000
    libemu: Standardise echo functions

    - There are three functions providing message printing for emu:
      emu_echo(), emu_warning() and emu_error().

Date:   Thu Dec 6 20:08:54 2012 +0000
    Tidy use of emu_echo and emu_error

    - Messages begin with lowercase letter.

    - Message do not end in a full stop or other punctuation.

Date:   Thu Dec 6 20:05:54 2012 +0000
    emu: Fix error output

Date:   Thu Dec 6 20:05:27 2012 +0000
    Add EMU_ECHO_PREFIX constant

Date:   Thu Dec 6 17:06:26 2012 +0000
    libemu: Add emu_echo() and emu_error() functions

Date:   Thu Dec 6 17:06:12 2012 +0000
    Fix makefile lib target location

Date:   Thu Dec 6 14:12:38 2012 +0000
    libemu: Refactor with 'local' variables

Date:   Thu Dec 6 14:10:05 2012 +0000
    libemu: Adds exit_command_not_found()

Date:   Fri Nov 30 15:59:16 2012 +0000
    libemu: Add emu_echo() and emu_error()

Date:   Mon Nov 26 22:54:10 2012 +0000
    Fix Makefile rules

Date:   Sun Nov 25 21:38:18 2012 +0000
    Add make-release

Date:   Sun Nov 25 20:34:52 2012 +0000
    sink: Update man page

Date:   Sun Nov 25 20:34:43 2012 +0000
    libemu: Add unit tests

Date:   Sun Nov 25 20:34:38 2012 +0000
    sink: Add unit tests

Date:   Sun Nov 25 20:34:25 2012 +0000
    test: Add extra assertions

Date:   Sun Nov 25 20:33:57 2012 +0000
    libemu: Fix permissions test

Date:   Sun Nov 25 20:33:28 2012 +0000
    sink: Fix sinks list

Date:   Sun Nov 25 20:33:12 2012 +0000
    init: Fix command args

Date:   Sun Nov 25 16:25:19 2012 +0000
    libemu: Fix print_sink_details()

Date:   Sun Nov 25 15:17:44 2012 +0000
    sink: Add relative path support

Date:   Sun Nov 25 13:17:27 2012 +0000
    test: Use subdirectories for unit tests

Date:   Sun Nov 25 13:10:29 2012 +0000
    libemu: Add exit_if_no_dir_permissions()

Date:   Sun Nov 25 13:05:11 2012 +0000
    sink: Add man page

Date:   Sun Nov 25 12:51:14 2012 +0000
    emu: Add hook man page description

Date:   Sun Nov 25 12:51:05 2012 +0000
    reset: Fix man page whitespace

Date:   Sun Nov 25 12:50:54 2012 +0000
    restore: Add man page

Date:   Sun Nov 25 03:05:44 2012 +0000
    log: Fix man page

Date:   Sun Nov 25 03:05:32 2012 +0000
    reset: Add man page

Date:   Sun Nov 25 02:58:03 2012 +0000
    log: Add man page

Date:   Sun Nov 25 02:53:00 2012 +0000
    snapshot: Add man page

Date:   Sun Nov 25 02:52:45 2012 +0000
    init: Fix man page

Date:   Sun Nov 25 02:52:21 2012 +0000
    emu: Use man page

Date:   Sun Nov 25 02:21:06 2012 +0000
    init: Use man page documentation

Date:   Sun Nov 25 02:20:42 2012 +0000
    init: Use ENV variable

Date:   Sun Nov 25 01:57:25 2012 +0000
    Change /usr/share to /usr/local/share

Date:   Sun Nov 25 01:51:11 2012 +0000
    init: Add permissions check/test

Date:   Sun Nov 25 01:49:59 2012 +0000
    Add Makefile doc rule

Date:   Sun Nov 25 01:48:36 2012 +0000
    init: Implement verbosity

Date:   Sun Nov 25 00:55:13 2012 +0000
    Changed Makefile verbosity

Date:   Sun Nov 25 00:37:55 2012 +0000
    init: Implement unit tests

Date:   Sun Nov 25 00:36:05 2012 +0000
    init: Implement command args

Date:   Sat Nov 24 21:40:34 2012 +0000
    Add test-suite and Makefile rule

    - '$ make check' executes the test suite.

    - Tests are individual executables in the test/ directory.

    - Tests should be as verbose as possible, with a 'running-commentary' of
      progress printed to stdout and errors printed to stderr.

    - If a test fails, it should exit as soon as possible ('set -e' in bash
      scripts).

    - test/run-tests.sh executes the tests and prints a summary of the
      results.

Date:   Sat Nov 24 17:24:21 2012 +0000
    Change directory layout

    - emu-util has been renamed libemu
    - use /usr/local/lib for libemu

Date:   Sat Nov 24 17:09:07 2012 +0000
    restore: fixed stdout

Date:   Sat Nov 24 16:50:02 2012 +0000
    snapshot: just some fixes

Date:   Sat Nov 24 16:49:40 2012 +0000
    restore: fix help

Date:   Sat Nov 24 16:49:29 2012 +0000
    emu: add 'clean' description

Date:   Sat Nov 24 16:33:04 2012 +0000
    snapshot: Change stdout

Date:   Sat Nov 24 16:32:53 2012 +0000
    reset: First implementation

Date:   Sat Nov 24 16:32:42 2012 +0000

emu/emu-push.sh
#!/bin/bash

source "$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )/../libexec/emu/libemu"
EMU_ECHO_PREFIX=$(basename "$0")

# Retrieve all stack files
#  @return STACK_FILES List of paths to stack files
get_all_stacks ()
{
    STACK_FILES="$(find $SOURCE_DIR/$EMU_DIR/stacks/ | tail -n+2)"

    if [[ "$STACK_FILES" == "" ]]
    then
        emu_error "no stacks! Add one with 'emu stack add <name> <path>'"
        emu_panic
        exit $EMU_EXIT_INCORRECT_COMMAND
    fi

    # check for all sync files
    for stack_file in $STACK_FILES
    do
        local stack="$(cat $stack_file)"
        if [[ ! -d "$stack/$EMU_DIR" ]]
        then
            emu_error "stack does not exist:\n  $stack\n"
            emu_panic
            exit $EMU_EXIT_ERROR
        fi
    done
}

# convert a list of stacks to stack files
#  @param  $@         List of stacks
#  @return STACK_FILES List of stack files paths
get_stacks ()
{
    STACK_FILES[$#]=$@
    for arg in $@
    do
        STACK_FILE="$(find $SOURCE_DIR/$EMU_DIR/stacks/ -name $arg)"
        if [[ "$STACK_FILE" == "" ]]
        then
            emu_error "stack '$arg' not found!"
            emu_panic
            exit $EMU_EXIT_INCORRECT_COMMAND
        fi
        STACK_FILES="$STACK_FILES\n$STACK_FILE"
    done
    STACK_FILES="$(echo -e $STACK_FILES | tail -n+2)"
}

# generate a snapshot hash
#  @param  $1   Path to stack directory
#  @return HASH A 40 digit snapshot UID
generate_hash ()
{
    HASH=$(printf '%x\n' $(date +'%s') | tail -c9)
    HASH=$HASH$(hash_directory "$1/$EMU_DIR/trees/new")

    # check if hash exists, and if so, wait
    if [[ -f "$1/$EMU_DIR/nodes/$HASH" ]]
    then
        sleep 1
        generate_hash "$1"
    fi
}

# generate a snapshot name
#  @param  $1       Path to stack directory
#  @var    HASH     The 40 digit snapshot UID
#  @return SNAPSHOT A snapshot name
generate_snapshot_name ()
{
    SNAPSHOT=$(date -d @$(printf '%d' 0x${HASH:0:8}) +'%Y-%m-%d %H.%M.%S')
}

# generate a snapshot node file
#  @param $1    Path to stack directory
#  @var   $HEAD The UID of the parent snapshot
#  @var   $HASH The 40 digit snapshot UID
generate_node ()
{
    local node="$1/$EMU_DIR/nodes/$HASH"
    touch "$node"
    cat <<EOF >> "$node"
Snapshot $SNAPSHOT
Parent   $HEAD
Date     $(date)
Source   $(cat $1/$EMU_DIR/SOURCE)
Size     $(du -sh "$1/$EMU_DIR/trees/$HASH" | sed -r 's/([0-9a-zA-Z]+)\s.*/\1/')
EOF
    chmod 0640 "$node"
}

# fetch the head of the stack tree, if there is one
#  @param  $1         Path to stack directory
#  @return HEAD       The UID of the HEAD snapshot
#  @return RSYNC_LINK The rsync link command, pointed to head
get_rsync_link ()
{
    HEAD="$(cat $1/$EMU_DIR/HEAD)"
    if [[ "$HEAD" != "" ]] && [[ -d "$1/$EMU_DIR/trees/$HEAD" ]]
    then
        RSYNC_LINK="--link-dest=$1/$EMU_DIR/trees/$HEAD"
    else
        RSYNC_LINK=""
    fi
}

# push the latest snapshot to HEAD
#  @param $1   Path to stack directory
#  @var   HASH The UID of the latest snapshot
push_head ()
{
    rm -f "$1/$EMU_DIR/HEAD"
    echo "$HASH" > "$1/$EMU_DIR/HEAD"
}

# fetch the parent of the current snapshot
#  @param  $1 Path to stack directory
#  @param  $2 Starting hash
#  @return    Next hash, or empty string if NULL
get_snapshot_parent ()
{
    echo "$(cat $1/$EMU_DIR/nodes/$2 2>/dev/null | grep Parent | sed -r 's/Parent +//')"
}

# Fetch the child of the current snapshot
#  @param  $1            Path to stack directory
#  @param  $2            Hash of starting point
#  @param  $3            Target hash
#  @return SNAPSHOT_HASH The hash of the snapshot's child
get_snapshot_child ()
{
    local next_snapshot_hash="$2"
    while [[ "$next_snapshot_hash" != "$3" ]]
    do
        SNAPSHOT_HASH="$next_snapshot_hash"
        next_snapshot_hash="$(get_snapshot_parent $1 $SNAPSHOT_HASH)"
    done
}

# deletes a snapshot
#  @param $1 path to stack directory
#  @param $2 snapshot to delete
#  @param $3 name of stack
remove_snapshot ()
{
    local stack_name="$3"

    get_snapshot_child "$1" "$(cat $1/$EMU_DIR/HEAD)" "$2"
    cat "$1/$EMU_DIR/nodes/$SNAPSHOT_HASH" | sed -r 's/^(Parent\s+).*/\1/i' > "$1/.emu/nodes/$SNAPSHOT_HASH.tmp"
    mv -f "$1/.emu/nodes/$SNAPSHOT_HASH.tmp" "$1/.emu/nodes/$SNAPSHOT_HASH"
    emu_echo "$stack_name: removing '$2'"
    rm -f $EMU_VERBOSE "$1/$EMU_DIR/nodes/$2"
    rm -f $EMU_VERBOSE "$1/$EMU_DIR/nodes/$2.MSG"
    rm -rf $EMU_VERBOSE "$1/$EMU_DIR/trees/$2"
    rm -rf $EMU_VERBOSE "$1/$(date -d @$(printf '%d' 0x5${2:0:8}) +'%Y-%m-%d %T')"
}

# remove the oldest in tree in a set of snapshots
#  @param $1 Path to stack directory
#  @param $2 name of the stack
pop_tail ()
{
    local stack_dir="$1"
    local stack_name="$2"

    # Get the root snapshot (one with no child).
    get_snapshot_child "$stack_dir" "$(cat $stack_dir/$EMU_DIR/HEAD)"
    remove_snapshot "$stack_dir" "$SNAPSHOT_HASH" "$stack_name"
}

# push the latest snapshot to HEAD
#  @param $1 Path to stack directory
#  @param $2 name of the stack
pop_tails ()
{
    local stack_dir="$1"
    local stack_name="$2"

    while [[ $(stack_snapshot_count "$stack_dir") -gt $(stack_max_snapshots "$stack_dir") ]]
    do
        next_snapshot_hash="$(cat $stack_dir/$EMU_DIR/HEAD)"
        pop_tail "$stack_dir" "$stack_name"
    done
}

# push the a snapshot on stack
#  @param  $1 Path to stack directory
#  @param  $2 Name of stack
#  @return    Integer success value (non-zero indicates failure)
push_snapshot_on_stack ()
{
    local stack_dir="$1"
    local stack_name="$2"

    lock_dir "$stack_dir"
    get_rsync_link "$stack_dir"

    if [ $EMU_DRY_RUN ]
    then
        rsync_dry_run="--dry-run"
    else
        pop_tails "$stack_dir" "$stack_name"
    fi

    emu_echo "$stack_name: pushing snapshot ($(stack_snapshot_count $stack_dir) of $(stack_max_snapshots $stack_dir))"

    rsync \
        $EMU_VERBOSE \
        $EMU_PROGRESS \
        $rsync_dry_run \
        --delete \
        --delete-excluded \
        --archive \
        --human-readable \
        --exclude="$SOURCE_DIR/$EMU_DIR/*" \
        --exclude-from="$SOURCE_DIR/$EMU_DIR/excludes" \
        "$RSYNC_LINK" \
        "$SOURCE_DIR/" "$stack_dir/$EMU_DIR/trees/new"
    if (( $? ))
    then
        unlock_dir "$stack_dir"
        return 1
    fi

    generate_hash "$stack_dir"
    generate_snapshot_name "$stack_dir"
    if [ ! $EMU_DRY_RUN ]
    then
        mv "$stack_dir/$EMU_DIR/trees/new" "$stack_dir/$SNAPSHOT"
        ln -s $EMU_VERBOSE "$stack_dir/$SNAPSHOT" "$stack_dir/$EMU_DIR/trees/$HASH"
        generate_node "$stack_dir"
        push_head "$stack_dir"

        if [ -n "$EMU_SNAPSHOT_MESSAGE" ]
        then
            echo "$EMU_SNAPSHOT_MESSAGE" > "$stack_dir/$EMU_DIR/nodes/$HASH.MSG"
        fi
    fi

    unlock_dir "$stack_dir"
    emu_echo "$stack_name: new snapshot '$SNAPSHOT'"
    emu_echo "$stack_name: HEAD at $HASH"

    test -n "$EMU_VERIFY_SNAPSHOT" && emu verify "$stack_name:$HASH"

    return 0
}

# push snapshots to all stack files
#  @var $STACK_FILES List of paths to stack files
push_snapshots ()
{
    local snapshot_failed=0

    for stack_file in $STACK_FILES
    do
        local stack="$(cat $stack_file)"
        local stack_name="$(basename $stack_file)"

        push_snapshot_on_stack "$stack" "$stack_name"
        if (( $? ))
        then
            snapshot_failed=1
            emu_error "snapshot to '$stack_name' failed"
        fi
    done

    return $snapshot_failed
}

main ()
{
    test -n "$EMU_DEBUG" && set -x
    exit_on_help_version $@

    while getopts ":nMvpfFV" OPT
    do
        case $OPT in
            n)
                EMU_DRY_RUN=1
                ;;
            M)
                if [ -z $EDITOR ]
                then
                    emu_error "EDITOR variable not specified"
                    exit $EMU_EXIT_INCORRECT_COMMAND
                fi

                $EDITOR EMU_SNAPSHOT_MSG
                EMU_SNAPSHOT_MESSAGE=$(cat EMU_SNAPSHOT_MSG)
                rm -f EMU_SNAPSHOT_MSG
                ;;
            v)
                EMU_VERBOSE="-v"
                ;;
            p)
                EMU_VERBOSE="-v"
                EMU_PROGRESS="-ph"
                ;;
            f)
                EMU_LOCK_FORCE=1
                ;;
            F)
                EMU_LOCK_FORCE=1
                EMU_LOCK_KILL=1
                ;;
            V)
                EMU_VERIFY_SNAPSHOT=1
                ;;
            \?)
                emu_error "invalid option: -$OPTARG"
                exit 1
                ;;
            :)
                emu_error "option -$OPTARG requires an argument"
                exit 1
                ;;
        esac
    done

    # dispose of args
    while [ "${1:0:1}" = "-" ]
    do
        shift
    done

    get_source_dir_or_fail
    execute_hooks "pre" "$SOURCE_DIR"

    if [[ "$@" == "" ]]
    then
        get_all_stacks
    else
        get_stacks $@
    fi

    lock_dir "$SOURCE_DIR"

    push_snapshots
    if (( $? ))
    then
        emu_error "at least one snapshot failed"
        emu_panic
        exit $EMU_EXIT_ERROR
    fi

    unlock_dir "$SOURCE_DIR"
    execute_hooks "post" "$SOURCE_DIR"
}
main $@
